// protoc --proto_path=src --cpp_out= ProtoData.proto 

package fantasybit;

enum MyNameStatus 
{
    none=1;
    notavil=2;
    requested=5;
    found=10;
    transaction_sent=15;
    confirmed=20;
}

message MyFantasyName
{
    optional string name=1;
    optional MyNameStatus status=2;
}

message Secret
{
    required string private_key = 1;
    optional MyFantasyName myfantasyname=2;
    optional NameTrans nametran = 10;
} 	

message OutData {
	enum Type { MYFANTASYNAME = 1; SNAPSHOT = 2; HEARTBEAT=3; }
	required Type type = 1;

	optional MyFantasyName myfantasyname = 2;	
  	extensions 100 to max;
}

// Elsewhere...
//extend OutData {
//	optional int32 test = 100;	
//}

message InData {
	enum Type { MINENAME = 1; QUIT=2; HEARTBEAT=3; CONNECT=4; MAKE_BLOCK=5; NEWNAME=6; PROJ=7;DATA=8;}
	required Type type = 1;
	
	optional string data = 2;	
	optional string data2 = 3;	
	optional int32 num = 4;
	optional DataTransition data_trans=5;
  	extensions 100 to max;
}

// Elsewhere...
//extend InData {
//	optional int32 test = 100;	
//}


message NameProof 
{
	enum Type { POW=1; TWEET=2; ORACLE=3; }
	optional Type type = 1;
	extensions 200 to max;
}

message TweetProof
{ 
	extend NameProof {
		optional TweetProof tweet_proof = 201;
	}

	optional string tweet = 10;
}

enum TransType {
	NAME = 0;
	PROJECTION = 1;
	RESULT = 2;
	DATA = 3;	
}

message Transaction {
	optional int32 version = 1;
	optional TransType type = 2;
	
	extensions 100 to max;
}
	
message NameTrans {
	extend Transaction {
		optional NameTrans name_trans = 200;
	}
	
	optional string fantasy_name = 10;
	optional bytes public_key=20;
	optional NameProof proof = 30;

}


message SignedTransaction {
	optional Transaction trans = 10;
	optional bytes id = 20;
	optional bytes sig = 30;	
	optional string fantasy_name = 40;
}

//message BlockHeader {
//	optional int32 num = 10;
//	optional bytes prev_id = 20; 
//}

message BlockHeader {
	enum Type { NORMAL=1; DATA=2; }

	optional int32 version = 1;
	optional int32 num = 10;
	optional bytes prev_id = 20; 		
	optional int32 timestamp = 21;
	optional bytes generator_pk=30;
	optional bytes generating_sig=40; //hash(prev(generating_sig) + generator_pk)
	optional int32 basetarget=50;
	optional Type blocktype = 60;
	optional bytes transaction_id = 70;
}

message SignedBlockHeader {
	optional BlockHeader head = 10;
	optional string sig = 30;	
}

message Block {
	optional SignedBlockHeader signedhead = 10;
	repeated SignedTransaction signed_transactions = 20;


	extensions 1000 to max;
 }

message NodeRequest
{
	enum Type { HANDSHAKE=1; BLOCK_REQUEST=2; HIGHT_REQUEST=3; NAT_TEST=4;}
	optional Type type = 10;
	optional string myip=20;
	optional int32 num=30;
	optional string myhost=40;
}

message NodeReply
{
	optional int32 hight=10;
	repeated string ips=20;
}

////////////////////////////////////////////////////////////
// FantasyPlayerPoints used for projections and results   //
// does not contain team                                  //
// player + week is enough                                //
////////////////////////////////////////////////////////////
message FantasyPlayerPoints {
	optional uint32 season = 10;	
	optional uint32 week = 20;
	optional string playerid = 30;
	optional int32 points = 40;	
}

////////////////////////////////////////////////////////////
// ProjectionTrans used to make projections 
////////////////////////////////////////////////////////////
message ProjectionTrans {
	extend Transaction {
		optional ProjectionTrans proj_trans = 201;
	}

	optional FantasyPlayerPoints fpp = 10;
}



message Data {
	enum Type {
		SCHEDULE = 0;
		PLAYER = 1;	
		TEAM = 2;
		RESULT=3;
	}

	optional int32 version = 10;
	optional Type type = 20;
	
	extensions 100 to max;
}

//player is int32 
//team is int32 

////////////////////////////////////////////////////////////
// PlayerData.playerid will be assigned a starting from 1000
// and mapped to player data 
// playerID 1-1000 is a teamid and is reserved for fantasy defense 
// when a new player is added we increment number 
////////////////////////////////////////////////////////////
message PlayerData {
	extend Data {
		optional PlayerData player_data = 101;
	}

	optional string playerid=10;
	optional string teamid=20;
}

////////////////////////////////////////////////////////////
// TeamData.teamid will be assigned a starting from 1 
// and mapped to team data
// when a new team is added to NFL - we will increment number 
// team 0 is a freeagent 
////////////////////////////////////////////////////////////
message TeamData {
	extend Data {
		optional TeamData team_data = 111;
	}

	optional string teamid=10;
}

message ResultData {
	extend Data {
		optional ResultData result_data = 202;
	}

	optional FantasyPlayerPoints fpp = 10;
}


message DataTransition {
	extend Transaction {
		optional DataTransition data_trans = 203;
	}

	enum Type { 
			ROSTER=1; //when cutdown to 53 man roster 
			SEASONSTART=2; // before week1 - week1 projection can start
			SEASONEND=3; //after last week of season 
			DRAFTOVER=4; //after nfl draft
			HEARTBEAT=5; //anytime
			GAMESTART=6; //list of teamids to cut off projections for week
			WEEKOVER=7; //clear all projections
	}

	enum State { 
		PREDRAFT = 1;
		PRESEASON = 2;
		ROSTER53MAN = 3;
		INSEASON = 4;
	}

	optional Type type =1;
	optional uint32 season =10;
	optional uint32 week = 20;
	repeated string teamid = 30;
	repeated Data data = 40;
}

message GlobalState {
	enum State { 
		PREDRAFT = 1;
		PRESEASON = 2;
		ROSTER53MAN = 3;
		INSEASON = 4;
	}

	optional State state =1;
	optional uint32 season =10;
}

message TeamState {
	enum State { 
		PREGAME = 1;
		INGAME = 2;
	}

	optional State state =1;
	optional uint32 week =10;
}